### PR TITLE
Fix AOT Compiler Compatibility Issues

### LINKED ISSUE #946: Maintain Xamarin iOS support by keeping the library "AOT-proof"
## Description
Xamarin iOS users need to be able to "seed" open-generic library API-s with the help of [AotCompilerTools](https://github.com/SixLabors/ImageSharp/blob/master/src/ImageSharp/Advanced/AotCompilerTools.cs), otherwise it will lead to [runtime exceptions](https://github.com/SixLabors/ImageSharp/issues/776#issuecomment-510424553) on that platform.

**Related:**  #767, #785, #776, #800

## Problems
Currently we are struggling to maintain this feature because of the following reasons:
- There is no one in the core team having experience with Xamarin iOS or any other AOT platform.
- We don't have any regression tests for this, we never know when a change breaks the "AOT-proofness" of an existing API
- We have introduced a massive change with #904, which will most likely break AOT & Xamarin iOS for many existing API-s in RC1. I have defined a point in epic #907 to investigate this, however in practice it turned out that I hardly have the resources to do so during this summer. (Neither I think @JimBobSquarePants can make it.) Therefore it will most likely not happen for RC1.

## Conclusion
With our current prioritites and resources, the only way to maintain this feature is community support. If you can provide any help (add missing AOT seeds, or fix regressions), feel free to open a Pull Request!

#### Long term solution
If someone could define a mono AOT test configuration for appveyor, that would be ace.

/CC @vicfergar  @davilovick  @gameleon-dev @Lapinou42 @dmanning23 @swoolcock

**EDIT**
**PS:** Could someone do a favor and chack the latest MyGet dev builds from on Xamarin iOS? (JPEG, PNG, Resize, Quantization etc.)

### PR DESCRIPTION
**Pull request by JimBobSquarePants**
### Prerequisites

- [x] I have written a descriptive pull-request title
- [x] I have verified that there are no overlapping [pull-requests](https://github.com/SixLabors/ImageSharp/pulls) open
- [x] I have verified that I am following matches the existing coding patterns and practice as demonstrated in the repository. These follow strict Stylecop rules :cop:.
- [x] I have provided test coverage for my change (where applicable)

### Description

Includes the work by @UltraNamahage to fix the AOT compiler issues when using the library with Xamarin iOS and Unity Android IL2CPP. Fixes #946

### CODE REVIEW

**Comment by UltraNamahage** [comment]
I have just confirmed the milestone.
You may need to seed for webp and tiff.

**Comment by brianpopow** [comment]
> You may need to seed for webp and tiff.

I will do that once this is merged.

File: src/ImageSharp/Advanced/AotCompilerTools.cs [hunk 1]
```diff
@@ -25,180 +40,510 @@ namespace SixLabors.ImageSharp.Advanced
...
+        /// If you are running ImageSharp on iOS and try to call SaveAsGif, it will throw an exception:
+        /// "Attempting to JIT compile method... OctreeFrameQuantizer.ConstructPalette... while running in aot-only mode."
+        /// The reason this happens is the SaveAsGif method makes heavy use of generics, which are too confusing for the AoT
+        /// compiler used on Xamarin.iOS. It spins up the JIT compiler to try and figure it out, but that is an illegal op on
+        /// iOS so it bombs out.
+        /// If you are getting the above error, you need to call this method, which will pre-seed the AoT compiler with the
+        /// necessary methods to complete the SaveAsGif call. That's it, otherwise you should NEVER need this method!!!
+        /// </remarks>
+        [Preserve]
         private static void SeedEverything()
```
Comment (brianpopow) [inline src/ImageSharp/Advanced/AotCompilerTools.cs:59]: the only nitpick i have is, this method could be called something else like `SeedPixelImplementations` or `SeedPIxelTypes`

**Review by brianpopow — APPROVED** [review]

File: src/ImageSharp/Advanced/AotCompilerTools.cs (hunk 1 above)
Comment (JimBobSquarePants) [reply src/ImageSharp/Advanced/AotCompilerTools.cs:59]: Feel free to change it once we do the WebP/Tiff updates.

**Comment by mackayn** [comment]
@JimBobSquarePants I'm still getting this issue in release config with version 2.1.1 on iOS, Xamarin Forms 5.0.0.2.0.41

`
System.ExecutionEngineException: Attempting to JIT compile method 'void SixLabors.ImageSharp.Formats.Png.PngScanlineProcessor:ProcessRgbaScanline<SixLabors.ImageSharp.PixelFormats.Rgba32> (SixLabors.ImageSharp.Configuration,SixLabors.ImageSharp.Formats.Png.PngHeader&,System.ReadOnlySpan`1<byte>,System.Span`1<SixLabors.ImageSharp.PixelFormats.Rgba32>,int,int)' while running in aot-only mode. See https://docs.microsoft.com/xamarin/ios/internals/limitations for more information.
 at SixLabors.ImageSharp.Formats.Png.PngDecoderCore.ProcessDefilteredScanline[TPixel] (System.ReadOnlySpan`1[T] defilteredScanline, SixLabors.ImageSharp.ImageFrame`1[TPixel] pixels, SixLabors.ImageSharp.Formats.Png.PngMetadata pngMetadata) <0x1067dd280 + 0x00614> in <face1fef1e124017bd2bd47b1c654a67#5c93716510ec001c2ff02230372d65b5>:0
 at SixLabors.ImageSharp.Formats.Png.PngDecoderCore.DecodePixelData[TPixel] (System.IO.Compression.DeflateStream compressedStream, SixLabors.ImageSharp.ImageFrame`1[TPixel] image, SixLabors.ImageSharp.Formats.Png.PngMetadata pngMetadata) <0x1067dc920 + 0x00337> in <face1fef1e124017bd2bd47b1c654a67#5c93716510ec001c2ff02230372d65b5>:0
 at SixLabors.ImageSharp.Formats.Png.PngDecoderCore.ReadScanlines[TPixel] (SixLabors.ImageSharp.Formats.Png.PngChunk chunk, SixLabors.ImageSharp.ImageFrame`1[TPixel] image, SixLabors.ImageSharp.Formats.Png.PngMetadata pngMetadata) <0x1067dc6a0 + 0x001f7> in <face1fef1e124017bd2bd47b1c654a67#5c93716510ec001c2ff02230372d65b5>:0
 at SixLabors.ImageSharp.Formats.Png.PngDecoderCore.Decode[TPixel] (SixLabors.ImageSharp.IO.BufferedReadStream stream, System.Threading.CancellationToken cancellationToken) <0x1067db790 + 0x00537> in <face1fef1e124017bd2bd47b1c654a67#5c93716510ec001c2ff02230372d65b5>:0
 at SixLabors.ImageSharp.Formats.ImageDecoderUtilities.Decode[TPixel] (SixLabors.ImageSharp.Formats.IImageDecoderInternals decoder, SixLabors.ImageSharp.Configuration configuration, System.IO.Stream stream, System.Func`3[T1,T2,TResult] largeImageExceptionFactory) <0x107733960 + 0x0009b> 48466 in <face1fef1e124017bd2bd47b1c654a67#5c93716510ec001c2ff02230372d65b5>:0
 at SixLabors.ImageSharp.Formats.ImageDecoderUtilities.Decode[TPixel] (SixLabors.ImageSharp.Formats.IImageDecoderInternals decoder, SixLabors.ImageSharp.Configuration configuration, System.IO.Stream stream) <0x106b27740 + 0x00087> 21367 in <face1fef1e124017bd2bd47b1c654a67#5c93716510ec001c2ff02230372d65b5>:0
 at SixLabors.ImageSharp.Formats.Png.PngDecoder.Decode[TPixel] (SixLabors.ImageSharp.Configuration configuration, System.IO.Stream stream) <0x106809e10 + 0x00053> 10940 in <face1fef1e124017bd2bd47b1c654a67#5c93716510ec001c2ff02230372d65b5>:0
 at SixLabors.ImageSharp.Formats.Png.PngDecoder.Decode (SixLabors.ImageSharp.Configuration configuration, System.IO.Stream stream) <0x10667d7f0 + 0x0002f> in <face1fef1e124017bd2bd47b1c654a67#5c93716510ec001c2ff02230372d65b5>:0
 at SixLabors.ImageSharp.Image.Decode (System.IO.Stream stream, SixLabors.ImageSharp.Configuration config) <0x10647bc00 + 0x0009b> in <face1fef1e124017bd2bd47b1c654a67#5c93716510ec001c2ff02230372d65b5>:0
 at SixLabors.ImageSharp.Image+<>c__DisplayClass137_0.<Load>b__0 (System.IO.Stream s) <0x106487380 + 0x00027> in <face1fef1e124017bd2bd47b1c654a67#5c93716510ec001c2ff02230372d65b5>:0
 at SixLabors.ImageSharp.Image.WithSeekableStream[T] (SixLabors.ImageSharp.Configuration configuration, System.IO.Stream stream, System.Func`2[T,TResult] action) <0x1067f43b0 + 0x000e7> in <face1fef1e124017bd2bd47b1c654a67#5c93716510ec001c2ff02230372d65b5>:0
 at SixLabors.ImageSharp.Image.Load (SixLabors.ImageSharp.Configuration configuration, System.IO.Stream stream, SixLabors.ImageSharp.Formats.IImageFormat& format) <0x106480cb0 + 0x00167> in <face1fef1e124017bd2bd47b1c654a67#5c93716510ec001c2ff02230372d65b5>:0
 at SixLabors.ImageSharp.Image.Load (SixLabors.ImageSharp.Configuration configuration, System.String path, SixLabors.ImageSharp.Formats.IImageFormat& format) <0x10647e810 + 0x000b7> in <face1fef1e124017bd2bd47b1c654a67#5c93716510ec001c2ff02230372d65b5>:0
 at SixLabors.ImageSharp.Image.Load (SixLabors.ImageSharp.Configuration configuration, System.String path) <0x10647dc10 + 0x00023> in <face1fef1e124017bd2bd47b1c654a67#5c93716510ec001c2ff02230372d65b5>:0
 at SixLabors.ImageSharp.Image.Load (System.String path) <0x10647db70 + 0x0003b> in <face1fef1e124017bd2bd47b1c654a67#5c93716510ec001c2ff02230372d65b5>:0

`

**Comment by JimBobSquarePants** [comment]
Tell Microsoft. They’re the ones breaking things not us.

**Pull request merged**